app/services/gov_one_auth_service.rb

```diff
@@ -0,0 +1,125 @@
+#
+# - exchange an authorisation code for tokens (access and id)
+# - exchange an access token for user info
+# - decode an id token to get the user's gov one id
+#
+# @see https://docs.sign-in.service.gov.uk/
+class GovOneAuthService
+  # @return [Hash{Symbol => String}]
+  CALLBACKS = {
+    login: "#{Rails.application.config.service_url}/users/auth/openid_connect/callback",
+    logout: "#{Rails.application.config.service_url}/users/sign_out",
+  }.freeze
+
+  # @return [Hash{Symbol => String}]
+  ENDPOINTS = {
+    login: "#{Rails.application.config.gov_one_base_uri}/authorize",
+    logout: "#{Rails.application.config.gov_one_base_uri}/logout",
+    token: "#{Rails.application.config.gov_one_base_uri}/token",
+    userinfo: "#{Rails.application.config.gov_one_base_uri}/userinfo",
+    jwks: "#{Rails.application.config.gov_one_base_uri}/.well-known/jwks.json",
+  }.freeze
+
+  extend Dry::Initializer
+
+  option :code, Types::Strict::String
+
+  # POST /token
+  # @return [Hash]
+  def tokens
+    uri, http = build_http(ENDPOINTS[:token])
+    token_request = Net::HTTP::Post.new(uri.path, { 'Content-Type' => 'application/x-www-form-urlencoded' })
+    token_request.set_form_data(token_body)
+    token_response = response(token_request, http)
+
+    JSON.parse(token_response.body)
+  rescue StandardError => e
+    Rails.logger.error "GovOneAuthService.tokens: #{e.message}"
+    {}
+  end
+
+  # GET /userinfo
+  # @param access_token [String]
+  # @return [Hash]
+  def user_info(access_token)
+    uri, http = build_http(ENDPOINTS[:userinfo])
+    userinfo_request = Net::HTTP::Get.new(uri.path, { 'Authorization' => "Bearer #{access_token}" })
+    userinfo_response = response(userinfo_request, http)
+
+    JSON.parse(userinfo_response.body)
+  rescue StandardError => e
+    Rails.logger.error "GovOneAuthService.user_info: #{e.message}"
+    {}
+  end
+
+  # @param request [Net::HTTP::Get, Net::HTTP::Post]
+  # @param http [Net::HTTP]
+  # @return [Net::HTTPResponse]
+  def response(request, http)
+    http.request(request)
+  end
+
+  # @param request [Net::HTTP::Get, Net::HTTP::Post]
+  # @return [Hash]
+  def handle_request(request)
```

Comment:
The method `#handle_request` defined here is not used by any other code. Do you need it?